src/background/services/deduplicator.ts

```diff
@@ -0,0 +1,73 @@
+import { Logger } from '@/shared/logger'
+
+type AsyncFn<T> = (...args: any[]) => Promise<T>
+
+interface CacheEntry {
+  promise: Promise<any>
+}
+
+interface DedupeOptions {
+  cacheFnArgs: boolean
+}
+
+export class Deduplicator {
+  private cache: Map<string, CacheEntry> = new Map()
+  private readonly duration = 5000
+
+  constructor(private logger: Logger) {}
+
+  dedupe<T extends AsyncFn<any>>(
+    fn: T,
+    options: DedupeOptions = { cacheFnArgs: false }
+  ): T {
+    return ((...args: Parameters<T>): ReturnType<T> => {
+      const key = this.generateCacheKey(fn, args, options.cacheFnArgs)
```

Comment:
Should we make this cache get an option to caller instead? I think there can be cases when JSON.stringify all args could cause performance issues if called often. We could maybe use: `{key: () => fn.name}` as default.

`wait` should be an option to be passed too.

Reply:
`fn.name` is already used as the default, if we do not want to cache with the args.

I like the approach of having the `wait` as an option when calling dedupe, instead of having it as a private field in the class.